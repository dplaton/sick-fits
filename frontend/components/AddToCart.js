import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

import Error from './ErrorMessage';

const ADD_TO_CART_MUTATION = gql`
    mutation addToCart($id: ID!) {
        addToCart(id: $id) {
            id
            quantity
        }
    }
`;

class AddToCart extends React.Component {
    render() {
        const { id } = this.props;

        return (
            <Mutation mutation={ADD_TO_CART_MUTATION} variables={{ id }}>
                {(addToCart, {error})=>(
                    <button onClick={addToCart}>Add to cart ðŸ›’</button>
                )}
            </Mutation>
        );
    }
}

export default AddToCart;
