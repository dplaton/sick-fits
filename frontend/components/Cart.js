import React from 'react';
import CartStyles from './styles/CartStyles';
import SickButton from './styles/SickButton';
import CloseButton from './styles/CloseButton';
import Supreme from './styles/Supreme';

const Cart = () => {
    return(
        <CartStyles open>
            <header>
                <CloseButton title="close">&times;</CloseButton>
                <Supreme>Your shopping cart</Supreme>
                <p>You have __ items in cart</p>
            </header>
            <footer>
                <p>$10.10</p>
                <SickButton>Checkout</SickButton>
            </footer>
        </CartStyles>
    )
}

export default Cart;