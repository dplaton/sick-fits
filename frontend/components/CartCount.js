import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { TransitionGroup, CSSTransition } from "react-transition-group";

/*
 * We customize the classes of the component so we have a flipping effect   
*/
const AnimationStyles = styled.span`
    position: relative;
    .count {
        display: block;
        position: relative;
        transition: all .4s;
        backface-visibility: hidden;
    }
    .count-enter {
        transform: scale(4) rotateX(0.5turn);
    }
    .count-enter-active {
        transform: rotateX(0);
    }
    .count-exit {
        top:0;
        position:absolute;
        transform: rotateX(0);
    }
    .count-exit-active {
        transform: scale(4) rotateX(0.5turn);
    }
`;

const Dot = styled.div`
    background: ${props => props.theme.red};
    border-radius: 50%;
    color: white;
    font-weight: 100;
    font-feature-settings: "tnum";
    font-variant-numeric: tabular-nums;
    line-height: 2rem;
    min-width: 3rem;
    margin-left: 1rem;
    padding: 0.5rem;
`;

/*
 * Using the TransitionGroup and the CSSTransition allows us to add a nice animation to the "cart count" badge
 * What React is doing is basically unmounting a component with count = x and mounting another one with count=x+1. 
 * In this proces React exposes a bunch of CSS classes so we can style the components (both the unmounted one and the mounted one)
 * 
 */
const CartCount = ({ count }) => (
    <AnimationStyles>
        <TransitionGroup>
            <CSSTransition
                unmountOnExit
                className="count"
                classNames="count"
                timeout={{ enter: 400, exit: 400 }}
                key={count}
            >
                <Dot>{count}</Dot>
            </CSSTransition>
        </TransitionGroup>
    </AnimationStyles>
);

export default CartCount;
