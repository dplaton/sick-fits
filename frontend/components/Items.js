import React, {Component} from 'react'
import gql from 'graphql-tag'
import {Query} from 'react-apollo'

const ALL_ITEMS_QUERY = gql `
    query ALL_ITEMS_QUERY {
        items {
            id
            title
            price
            description
            image
            largeImage
        }
    }
`;

class Items extends Component {
    render() {
        return (
            <div>
                <Query query={ALL_ITEMS_QUERY}>
                    {
                        ({data, error, loading}) => {
                            if (loading) return <p>Loading...</p>
                            if (error) return <p>Error! {error}</p>

                            return <p>Hey, I found {data.items.length}</p>;
                        }
                    } 
                </Query>
            </div>
        )
    }
}

export default Items