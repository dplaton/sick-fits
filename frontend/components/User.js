import { Query } from "react-apollo";
import gql from "graphql-tag";
import PropTypes from "prop-types";

const CURRENT_USER_QUERY = gql`
    query {
        me {
            id
            name
            email
            permissions
            cart {
                id
                quantity
                item {
                    id
                    title
                    price
                    image
                    description
                }
            }
        }
    }
`;

// this is how we build our own render prop
const User = props => (
    // the children of this component must be functions which receive the 'payload' as argument
    <Query {...props} query={CURRENT_USER_QUERY}>
        {payload => props.children(payload)}
    </Query>
);

User.propTypes = {
    children: PropTypes.func.isRequired
};

export default User;
export { CURRENT_USER_QUERY };
