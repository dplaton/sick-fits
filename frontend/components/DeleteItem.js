import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';

import { ALL_ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id) {
            id
        }
    }
`;

class DeleteItem extends Component {
    update = (cache, payload) => {
        // manually update the cache on the client instead of re-fetching everything

        // 1. Read the items from the cache
        const data = cache.readQuery({ query: ALL_ITEMS_QUERY });

        // 2. Filter the deleted item
        data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
        cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
    };

    render() {
        return (
            // relevant docs: https://www.apollographql.com/docs/react/essentials/mutations.html
            // the "udpdate" prop specifies a function to be used as a 'callback' after the mutation was executed
            <Mutation mutation={DELETE_ITEM_MUTATION} variables={{ id: this.props.id }} update={this.update}>
                {(deleteItem, { error }) => (
                    <button
                        onClick={() => {
                            if (confirm('Are you sure you want to delete this item?')) {
                                deleteItem()
                                    .catch(err => { alert(err.message) });
                            }
                        }}>
                        {this.props.children}
                    </button>
                )}
            </Mutation>
        );
    }
}

export default DeleteItem;
