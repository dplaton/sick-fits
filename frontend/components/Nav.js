import Link from "next/link";
import NavStyles from "./styles/NavStyles";
import User from './User'

const Nav = () => (
    <NavStyles>
        <User>
            {({data: {me}}) => {
                // two level destructuring, like payload.data.me
                if (me) {
                    return <p>{me.name}</p>
                } else {
                    return <p>Nobody</p>
                }
            }}
        </User>
        <Link href="/items">
            <a>Shop</a>
        </Link>
        <Link href="/sell">
            <a>Sell</a>
        </Link>
        <Link href="/signup">
            <a>Signup</a>
        </Link>
        <Link href="/orders">
            <a>Orders</a>
        </Link>
        <Link href="/me">
            <a>Account</a>
        </Link>
    </NavStyles>
);

export default Nav;
