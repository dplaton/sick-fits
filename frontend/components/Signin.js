import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

import Form from "./styles/Form";
import Error from "./ErrorMessage";
import { CURRENT_USER_QUERY } from "./User";

const SIGNIN_MUTATION = gql`
    mutation SIGNIN_MUTATION($email: String!, $password: String!) {
        signIn(email: $email, password: $password) {
            id
            email
            name
        }
    }
`;

class Signin extends Component {
    state = {
        email: "",
        password: ""
    };

    saveToState = event => {
        this.setState({ [event.target.name]: event.target.value });
    };

    render() {
        return (
            <Mutation
                mutation={SIGNIN_MUTATION}
                variables={this.state}
                refetchQueries={[{ query: CURRENT_USER_QUERY }]}
            >
                {(signUp, { error, loading }) => {
                    return (
                        <Form
                            method="POST"
                            onSubmit={async e => {
                                e.preventDefault();
                                await signUp();
                                this.setState({
                                    name: "",
                                    email: "",
                                    password: ""
                                });
                            }}
                        >
                            <fieldset disabled={loading} aria-busy={loading}>
                                <Error error={error} />
                                <h2>Sign in to your account</h2>
                                <label htmlFor="email">
                                    Email
                                    <input
                                        type="email"
                                        name="email"
                                        placeholder="Your email"
                                        value={this.state.email}
                                        onChange={this.saveToState}
                                    />
                                </label>
                                <label htmlFor="password">
                                    Password
                                    <input
                                        type="password"
                                        name="password"
                                        placeholder="Your password"
                                        value={this.state.password}
                                        onChange={this.saveToState}
                                    />
                                </label>
                                <button type="submit">Signin</button>
                            </fieldset>
                        </Form>
                    );
                }}
            </Mutation>
        );
    }
}

export default Signin;
