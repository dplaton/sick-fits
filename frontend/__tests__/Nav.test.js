import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/User";
import { fakeUser } from "../lib/testUtils";
import Signout from "../components/Signout";
import Signin from "../components/Signin";

const notSignedInMocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: null } }
    }
];

const signedInMocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: fakeUser() } }
    }
];

describe("<Nav/>", () => {
    it("shows the required options when the user is logged in", async () => {
        
        const wrapper = mount(
            <MockedProvider mocks={signedInMocks}>
                <Nav />
            </MockedProvider>
        );

        await wait();
        wrapper.update()
        const nav = wrapper.find('[data-test="nav"]');
        expect(nav.contains(<Signout/>)).toBe(true);
    });

    it('shows the sign-in option when the user is not logged in', async () => {
 
        const wrapper = mount(
            <MockedProvider mocks={notSignedInMocks}>
                <Nav />
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        const nav = wrapper.find('[data-test="nav"]');
        expect(toJSON(nav)).toMatchSnapshot();
    });
});
