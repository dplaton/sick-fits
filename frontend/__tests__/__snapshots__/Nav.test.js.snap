// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> shows the correct number of cart entries 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-1 dUHOkR"
>
  9
</div>
`;

exports[`<Nav/> shows the sign-in option when the user is not logged in 1`] = `
<ul
  className="NavStyles-sc-11c0d2g-0 bjCNSW"
  data-test="nav"
>
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/signup"
  >
    <a
      href="/signup"
      onClick={[Function]}
    >
      Sign in
    </a>
  </Link>
</ul>
`;
